src/csharp/Grpc.Core.Api/ServerServiceDefinition.cs

```diff
@@ -62,7 +60,10 @@ public static Builder CreateBuilder()
         /// </summary>
         public class Builder
         {
-            readonly Dictionary<string, IServerCallHandler> callHandlers = new Dictionary<string, IServerCallHandler>();
+            // to maintain legacy behavior, we need to detect duplicate keys and throw the same exception as before
```

Comment:
I think this is a little redundant here. I think you can just make it a Dictionary<string, Action<ServiceBinderBase>> and then get the list of addMethodActions via dictionary.Keys which is read-only.

Reply:
+1 (tho it is dictionary.Values, not dictionary.Keys)

If you did this then it would want this to change to an `IEnumerable<Action<ServiceBinderBase>>` - https://github.com/grpc/grpc/pull/17889/files#r252833644

Reply:
Curious why the order needs to be preserved? The builder used to add to a `Dictionary` which is then stored in the `ServerServiceDefinition`. Then it is iterated over in Server.AddServiceDefinitionInternal which doesn't guarantee order.